zuko/bayesian.py

```diff
@@ -0,0 +1,367 @@
+                    continue
+                safe_name = name.replace(".", "_")
+
+                # Weights - always read from base model unless learn_means=True
+                if self.learn_means and safe_name + "_weight" in self.weight_means:
+                    w_mu = self.weight_means[safe_name + "_weight"]
+                else:
+                    w_mu = module.weight
+
+                # Apply same softclip as BayesianLinear and use sqrt of exp(logvar)
+                w_logvar = _softclip(self.weight_logvars[safe_name + "_weight"], 11)
+                sampled[name + ".weight"] = w_mu + torch.exp(w_logvar).sqrt() * torch.randn(
+                    w_mu.shape, generator=self.generator, device=w_mu.device
+                )
+
+                # Bias - always read from base model unless learn_means=True
+                if module.bias is not None:
+                    if self.learn_means and safe_name + "_bias" in self.bias_means:
+                        b_mu = self.bias_means[safe_name + "_bias"]
+                    else:
+                        b_mu = module.bias
+
+                    # Apply same softclip as BayesianLinear and use sqrt of exp(logvar)
+                    b_logvar = _softclip(self.bias_logvars[safe_name + "_bias"], 11)
+                    sampled[name + ".bias"] = b_mu + torch.exp(b_logvar).sqrt() * torch.randn(
+                        b_mu.shape, generator=self.generator, device=b_mu.device
+                    )
+
+        return sampled
+
+    @contextmanager
+    def sample(self, trick=False):
+        """Context manager yielding the model with reparametrized forward methods for linear model.
+        This is not cloning the model but just temporarily replacing the forward methods of the
+        selected linear layers to use the local reparameterization trick during training, or
+        sampling once during evaluation."""
+
+        if not trick and not self.training:
+            sampled_params = self._sample_params()
+            with torch.nn.utils.stateless._reparametrize_module(self.base, sampled_params):
+                yield self.base
+        else:
+            with self._reparameterization_context():
+                yield self.base
+
+    @contextmanager
+    def _reparameterization_context(self):
+        """Context manager that temporarily replaces linear layer forwards with reparameterized versions."""
+        # Store original forward methods and replace them
+        original_forwards = {}
+
+        # Sample a new seed for each module call
+        for name, module in self.base.named_modules():
+            if isinstance(module, (Linear, MaskedLinear)):
+                # Skip non-selected layers when a subset was provided
+                if self._requested_bayesian is not None and name not in self._requested_bayesian:
+                    continue
+
+                seed = int(torch.randint(0, 2**31 - 1, (1,), generator=self.generator).item())
```

Comment:
You can use the global random state here. 

```suggestion
                seed = torch.randint(0, 2**31 - 1, ()).item()
```